File: app/src/main/java/com/itba/runningMate/di/ProductionDependencyContainer.java
package com.itba.runningMate.di;

import android.content.Context;

import com.itba.runningMate.achievements.elements.AchievementsElementView;
import com.itba.runningMate.db.RunDao;
import com.itba.runningMate.db.RunDb;
import com.itba.runningMate.repository.achievementsstorage.AchievementsStorage;
import com.itba.runningMate.repository.runningstate.RunningStateStorage;
import com.itba.runningMate.repository.run.RunRepository;
import com.itba.runningMate.utils.providers.files.CacheFileProvider;
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider;

public class ProductionDependencyContainer implements DependencyContainer {

    private final Dependency dependency;

    private SchedulerProvider schedulerProvider;
    private CacheFileProvider cacheFileProvider;
    private RunningStateStorage runningStateStorage;
    private AchievementsStorage achievementsStorage;
    private RunRepository runRepository;
    private RunDb runDb;

    public ProductionDependencyContainer(final Context context) {
        dependency = new Dependency(context);
    }


    @Override
    public Context getApplicationContext() {
        return dependency.getApplicationContext();
    }

    @Override
    public SchedulerProvider getSchedulerProvider() {
        if (schedulerProvider == null) {
            schedulerProvider = dependency.provideSchedulerProvider();
        }
        return schedulerProvider;
    }

    @Override
    public CacheFileProvider getCacheFileProvider() {
        if (cacheFileProvider == null) {
            cacheFileProvider = dependency.provideCacheFileProvider();
        }
        return cacheFileProvider;
    }

    @Override
    public RunningStateStorage getRunningStateStorage() {
        if (runningStateStorage == null) {
            runningStateStorage = dependency.provideRunningStateStorage();
        }
        return runningStateStorage;
    }

    @Override
    public AchievementsStorage getAchievementsStorage() {
        if (achievementsStorage == null) {
            achievementsStorage = dependency.provideAchievementsStorage();
        }
        return achievementsStorage;
    }

    @Override
    public RunRepository getRunRepository() {
        if (runRepository == null) {
            runRepository = dependency.provideRunRepository(getRunDao(), getSchedulerProvider());
        }
        return runRepository;
    }

    private RunDao getRunDao() {
        if (runDb == null) {
            runDb = dependency.provideRunDb();
        }
        return runDb.RunDao();
    }

}

File: app/src/main/java/com/itba/runningMate/di/ProductionDependencyContainer.kt
package com.itba.runningMate.di

import android.content.Context
import com.itba.runningMate.db.achievement.AchievementDao
import com.itba.runningMate.db.achievement.AchievementDb
import com.itba.runningMate.db.run.RunDao
import com.itba.runningMate.db.run.RunDb
import com.itba.runningMate.repository.achievements.AchievementsRepository
import com.itba.runningMate.repository.aggregaterunmetrics.AggregateRunMetricsStorage
import com.itba.runningMate.repository.run.RunRepository
import com.itba.runningMate.repository.runningstate.RunningStateStorage
import com.itba.runningMate.services.location.TrackingLocationUpdatesDispatcher
import com.itba.runningMate.utils.providers.files.CacheFileProvider
import com.itba.runningMate.utils.providers.schedulers.SchedulerProvider

class ProductionDependencyContainer(context: Context) : DependencyContainer {

    private val dependency: Dependency = Dependency(context)

    private var trackingLocationUpdatesDispatcher: TrackingLocationUpdatesDispatcher? = null
    private var schedulerProvider: SchedulerProvider? = null
    private var cacheFileProvider: CacheFileProvider? = null
    private var runningStateStorage: RunningStateStorage? = null
    private var aggregateRunMetricsStorage: AggregateRunMetricsStorage? = null
    private var achievementsRepository: AchievementsRepository? = null
    private var runRepository: RunRepository? = null
    private var runDb: RunDb? = null
    private var achievementDb: AchievementDb? = null

    override fun getApplicationContext(): Context {
        return dependency.applicationContext
    }

    override fun getSchedulerProvider(): SchedulerProvider {
        if (schedulerProvider == null) {
            schedulerProvider = dependency.provideSchedulerProvider()
        }
        return schedulerProvider!!
    }

    override fun getCacheFileProvider(): CacheFileProvider {
        if (cacheFileProvider == null) {
            cacheFileProvider = dependency.provideCacheFileProvider()
        }
        return cacheFileProvider!!
    }

    override fun getRunningStateStorage(): RunningStateStorage {
        if (runningStateStorage == null) {
            runningStateStorage = dependency.provideRunningStateStorage()
        }
        return runningStateStorage!!
    }

    override fun getAchievementsRepository(): AchievementsRepository {
        if (achievementsRepository == null) {
            achievementsRepository = dependency.provideAchievementsRepository(getAchievementDao())
        }
        return achievementsRepository!!
    }

    override fun getAggregateRunMetricsStorage(): AggregateRunMetricsStorage {
        if (aggregateRunMetricsStorage == null) {
            aggregateRunMetricsStorage = dependency.provideAggregateRunMetricsStorage()
        }
        return aggregateRunMetricsStorage!!
    }

    override fun getRunRepository(): RunRepository {
        if (runRepository == null) {
            runRepository = dependency.provideRunRepository(getRunDao())
        }
        return runRepository!!
    }

    private fun getRunDao(): RunDao {
        if (runDb == null) {
            runDb = dependency.provideRunDb()
        }
        return runDb!!.RunDao()
    }

    private fun getAchievementDao(): AchievementDao {
        if (achievementDb == null) {
            achievementDb = dependency.provideAchievementsDb()
        }
        return achievementDb!!.AchievementDao()
    }

    override fun getTrackingLocationUpdatesDispatcher(): TrackingLocationUpdatesDispatcher {
        if (trackingLocationUpdatesDispatcher == null) {
            trackingLocationUpdatesDispatcher =
                dependency.provideTrackingLocationUpdatesDispatcher()
        }
        return trackingLocationUpdatesDispatcher!!
    }

}